### PR TITLE
OP-159 externalize jwt secrets and improve spring security

### PR DESCRIPTION
**Pull request by mwithi**
Preparing for release 1.12.0... see OP-159, paired with https://github.com/informatici/openhospital-core/pull/862.

- [x] Get rid of hardcoded secrets.
- [x] Get rid of deprecated methods.

### CODE REVIEW

File: src/main/java/org/isf/config/SecurityConfig.java [hunk 1]
```diff
@@ -287,9 +269,10 @@ protected void configure(final HttpSecurity http) throws Exception {
 				.logoutUrl("/auth/logout")
 				.addLogoutHandler(customLogoutHandler)
 				.permitAll();    
-        }
+        return http.build();
+    }
 
-    private JWTConfigurer securityConfigurerAdapter() {
+	private JWTConfigurer securityConfigurerAdapter() {
```
Comment (mwithi) [inline src/main/java/org/isf/config/SecurityConfig.java:275]:
```suggestion
    private JWTConfigurer securityConfigurerAdapter() {
```

File: src/main/java/org/isf/config/SpringFoxConfig.java [hunk 2]
```diff
@@ -49,11 +50,8 @@
 @EnableSwagger2
 public class SpringFoxConfig {
 
-	@Value("${api.host:localhost:8080}")
-	String host;
-
-	@Value("${api.protocol:http}")
-	String protocol;
+	@Autowired
+	private Environment env;
```
Comment (mwithi) [inline src/main/java/org/isf/config/SpringFoxConfig.java:54]:
```suggestion
    @Autowired
    private Environment env;
```

File: src/main/java/org/isf/security/jwt/JWTFilter.java [hunk 3]
```diff
@@ -53,7 +53,8 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
 			throws IOException, ServletException {
 		HttpServletRequest httpServletRequest = (HttpServletRequest) servletRequest;
 		String jwt = resolveToken(httpServletRequest);
-		if (StringUtils.hasText(jwt) && this.tokenProvider.validateToken(jwt)) {
+
+        if (StringUtils.hasText(jwt) && this.tokenProvider.validateToken(jwt) && !this.tokenProvider.isTokenExpired(jwt)) {
```
Comment (mwithi) [inline src/main/java/org/isf/security/jwt/JWTFilter.java:57]:
```suggestion
		if (StringUtils.hasText(jwt) && this.tokenProvider.validateToken(jwt) && !this.tokenProvider.isTokenExpired(jwt)) {
```

File: src/main/java/org/isf/security/jwt/TokenProvider.java [hunk 4]
```diff
@@ -40,42 +44,73 @@
...
+public class TokenProvider implements Serializable {
 
     private final Logger log = LoggerFactory.getLogger(TokenProvider.class);
-
-    private static final String AUTHORITIES_KEY = "auth";
+    
+	@Autowired
+	private Environment env;
+	
+	private static final String AUTHORITIES_KEY = "auth";
```
Comment (mwithi) [inline src/main/java/org/isf/security/jwt/TokenProvider.java:62]:
```suggestion
    @Autowired
    private Environment env;

    private static final String AUTHORITIES_KEY = "auth";
```

File: src/main/java/org/isf/security/jwt/TokenProvider.java [hunk 5]
```diff
@@ -108,14 +139,19 @@ public Authentication getAuthentication(String token) {
...
+			log.error("Invalid JWT token: {}", e.getMessage());
+		} catch (ExpiredJwtException e) {
+			log.error("JWT token is expired: {}", e.getMessage());
+		} catch (UnsupportedJwtException e) {
+			log.error("JWT token is unsupported: {}", e.getMessage());
+		} catch (IllegalArgumentException e) {
+			log.error("JWT claims string is empty: {}", e.getMessage());
+		}
+		return false;
+	}
```
Comment (mwithi) [inline src/main/java/org/isf/security/jwt/TokenProvider.java:156]:
```suggestion
    public boolean validateToken(String token) {
        try {
            Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(token);
            return true;
        } catch (MalformedJwtException e) {
            log.error("Invalid JWT token: {}", e.getMessage());
        } catch (ExpiredJwtException e) {
            log.error("JWT token is expired: {}", e.getMessage());
        } catch (UnsupportedJwtException e) {
            log.error("JWT token is unsupported: {}", e.getMessage());
        } catch (IllegalArgumentException e) {
            log.error("JWT claims string is empty: {}", e.getMessage());
        }
        return false;
    }
```

File: README.md [hunk 6]
```diff
@@ -20,11 +20,12 @@ For the moment, to build this project you should
         cd openhospital-api
         mvn clean install -DskipTests=true
         
- 3. prepare settings from rsc/*.dist files
+ 3. prepare settings from each rsc/*.dist files
```
Comment (dbmalkovsky) [inline README.md:23]:
```suggestion
 3. prepare settings from each rsc/*.dist file
```

File: src/main/java/org/isf/login/rest/LoginController.java [hunk 7]
```diff
@@ -57,23 +55,20 @@
...
 	private HttpSession httpSession;
 
 	@Autowired
 	private SessionAuditManager sessionAuditManager;
+	
+	@Autowired
+	private TokenProvider tokenProvider;
+	
+	@Autowired
+    private CustomAuthenticationManager authenticationManager;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/login/rest/LoginController.java:68]:
```suggestion
        private CustomAuthenticationManager authenticationManager;
```

File: src/main/java/org/isf/security/CustomAuthenticationManager.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
+package org.isf.security;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/security/CustomAuthenticationManager.java:1]: Missing license header

**Review by dbmalkovsky — APPROVED** [review]
LGTM (with my limited knowledge in this area) after a few small cosmetic changes.

File: src/main/java/org/isf/security/CustomAuthenticationManager.java (hunk 8 above)
Comment (mwithi) [reply src/main/java/org/isf/security/CustomAuthenticationManager.java:1]: done, thanks!

**Review by averni — APPROVED** [review]
LGTM

**Comment by goto-eof** [comment]
For me is ok

**Comment by mwithi** [comment]
Thank you guys, I'm gonna merge all this stuff, starting from core + api + gui

**Pull request merged**